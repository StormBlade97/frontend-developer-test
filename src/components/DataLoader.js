import React from 'react';
import { FETCH_STATE } from '../const';

import Box from '@material-ui/core/Box';
import DataTable from './DataTable';
import TableControl from './TableControl';
import { format } from 'date-fns';
import { sortByDate, mapToTableData, uniqueBy } from './utils';

class DataLoader extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            data: [],
            sortDir: 'asc',
            errorMessage: '',
            fetchState: FETCH_STATE.INITIAL
        }
    }
    
    handleFetchData = async () => {
        try {
            this.setState({ ...this.state, fetchState: FETCH_STATE.PENDING });
            const result = await this.props.api();
            
            this.setState({
                ...this.state, fetchState: FETCH_STATE.SUCCESSFUL,
                data: uniqueBy([...this.state.data, ...result.data], 'timestamp')
            });
        }
        catch (error) {
            this.setState({ ...this.state, fetchState: FETCH_STATE.ERROR, errorMessage: error.message});
        }
    }

    toggleSortDir = async () => {
        this.setState({
            ...this.state,
            sortDir: this.state.sortDir === 'asc' ? 'desc' : 'asc'
        });
    }

    componentDidMount() {
        this.handleFetchData();
    }

    render() {
        const { sortDir, data, errorMessage, fetchState } = this.state;
        const sortFn = sortByDate(this.state.sortDir);
        const rows = mapToTableData(data)
            .sort(sortFn)
            .map(({ date, ...rest }) => ({ ...rest, date: format(date, 'yyyy-MM-dd') }));
    
        return (
            <Box data-test="app-box" m={2}>
                <DataTable
                    errorMessage={errorMessage}
                    fetchState={fetchState}
                    columns={['Date', 'User id', 'Old value', 'New value']}
                    sortBy={'Date'}
                    sortDirection={sortDir}
                    toggleSort={this.toggleSortDir}
                    rows={rows}
                    tableController={
                        <TableControl
                            data-test='table-controller'
                            fetchState={fetchState}
                            fetchData={this.handleFetchData} />
                    }
                />
            </Box>
        );
    }
}
export default DataLoader;