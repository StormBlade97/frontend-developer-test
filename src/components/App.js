import React from 'react';
import api from '../lib/api';
import Container from '@material-ui/core/Container';
import DataLoader from './DataLoader';
export const App = () => {
  return (
    <Container className="app" fixed>
      <DataLoader key={1} api={api.getUsersDiff}></DataLoader>
      <DataLoader key={2} api={api.getProjectsDiff}></DataLoader>
    </Container>
  );
};

export default App;
