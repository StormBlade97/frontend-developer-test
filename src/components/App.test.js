import React from 'react';
import { shallow } from 'enzyme';
import { App } from './App';
import DataLoader from './DataLoader';

describe('<App />', () => {
  describe('renders', () => {
    it('renders', () => {
      expect(shallow(<App />).getElement()).toMatchSnapshot();
    })
    it('renders two DataLoader', () => {
      expect(shallow(<App />).find(DataLoader).length).toBe(2);
    })
  });
});
