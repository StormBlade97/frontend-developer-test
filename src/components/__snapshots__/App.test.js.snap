// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders renders 1`] = `
<ForwardRef(WithStyles)
  className="app"
  fixed={true}
>
  <DataLoader
    api={[Function]}
  />
  <DataLoader
    api={[Function]}
  />
</ForwardRef(WithStyles)>
`;
